crates/weaver_common/src/diagnostic.rs

```diff
@@ -139,6 +151,43 @@ impl DiagnosticMessages {
     }
 }
 
+/// An extension trait for `Result` that captures the diagnostic messages
+pub trait ResultExt<T, E> {
+    /// Captures the diagnostic messages into the provided `DiagnosticMessages`
+    /// or returns the value if there are no diagnostic messages.
+    fn capture_diag_msgs_into(self, diags: &mut DiagnosticMessages) -> Option<T>;
```

Comment:
I like these additions.  Will need to figure out if we have the right set of combinators here.